service/history/historybuilder/history_builder.go

```diff
@@ -0,0 +1,813 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package historybuilder
+
+import (
+	"time"
+
+	commandpb "go.temporal.io/api/command/v1"
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	failurepb "go.temporal.io/api/failure/v1"
+	historypb "go.temporal.io/api/history/v1"
+	sdkpb "go.temporal.io/api/sdk/v1"
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	updatepb "go.temporal.io/api/update/v1"
+	workflowpb "go.temporal.io/api/workflow/v1"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common/clock"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+)
+
+const (
+	HistoryBuilderStateMutable HistoryBuilderState = iota
+	HistoryBuilderStateImmutable
+	HistoryBuilderStateSealed
+)
+
+type (
+	EventFactory struct {
+		timeSource clock.TimeSource
+		version    int64
+	}
+
+	HistoryBuilder struct {
+		EventStore
+		EventFactory
+	}
+
+	HistoryBuilderState int
+
+	HistoryMutation struct {
+		// events to be persisted to events table
+		DBEventsBatches [][]*historypb.HistoryEvent
+		// events to be buffer in execution table
+		DBBufferBatch []*historypb.HistoryEvent
+		// whether to clear buffer events on DB
+		DBClearBuffer bool
+		// accumulated buffered events, equal to all buffer events from execution table
+		MemBufferBatch []*historypb.HistoryEvent
+		// scheduled to started event ID mapping for flushed buffered event
+		ScheduledIDToStartedID map[int64]int64
+	}
+
+	TaskIDGenerator func(number int) ([]int64, error)
+
+	BufferedEventFilter func(*historypb.HistoryEvent) bool
+)
+
+func NewMutableHistoryBuilder(
```

Comment:
Consider call this just `NewMutable`. On caller side it will be `historybuilder.NewMutable`. Also I would remove `HistoryBuilder` from all other names in this file besides `HistoryBuilder struct` itself, of course.